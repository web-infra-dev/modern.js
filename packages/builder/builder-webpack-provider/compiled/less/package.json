{"name":"less","author":{"name":"Alexis Sellier","email":"self@cloudhead.net"},"version":"4.1.3","license":"Apache-2.0"}
