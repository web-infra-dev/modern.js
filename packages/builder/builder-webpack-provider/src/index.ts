export { builderWebpackProvider } from './provider';
export type { BuilderWebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export type { WebpackBuildError } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  BuilderPlugin,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CSSExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
