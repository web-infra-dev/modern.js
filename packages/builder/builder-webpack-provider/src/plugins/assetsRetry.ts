import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
