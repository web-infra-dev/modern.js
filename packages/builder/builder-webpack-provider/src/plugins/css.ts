import path from 'path';
import assert from 'assert';
import {
  CSS_REGEX,
  isUseCssSourceMap,
  isLooseCssModules,
  getBrowserslistWithDefault,
  type BuilderTarget,
  type BuilderContext,
} from '@modern-js/builder-shared';
import _, { merge as deepMerge } from '@modern-js/utils/lodash';
import type {
  WebpackChain,
  BuilderPlugin,
  CSSExtractOptions,
  CSSLoaderOptions,
  NormalizedConfig,
  StyleLoaderOptions,
  ModifyWebpackChainUtils,
} from '../types';
import type { AcceptedPlugin, ProcessOptions } from 'postcss';
import { getCssnanoDefaultOptions } from './minimize';

export const isUseCssExtract = (
  config: NormalizedConfig,
  target: BuilderTarget,
) =>
  !config.output.disableCssExtract &&
  target !== 'node' &&
  target !== 'web-worker';

// If the target is 'node' or 'web-worker' and the modules option of css-loader is enabled,
// we must enable exportOnlyLocals to only exports the modules identifier mappings.
// Otherwise, the compiled CSS code may contain invalid code, such as `new URL`.
// https://github.com/webpack-contrib/css-loader#exportonlylocals
export const normalizeCssLoaderOptions = (
  options: CSSLoaderOptions,
  exportOnlyLocals: boolean,
) => {
  if (options.modules && exportOnlyLocals) {
    let { modules } = options;
    if (modules === true) {
      modules = { exportOnlyLocals: true };
    } else if (typeof modules === 'string') {
      modules = { mode: modules, exportOnlyLocals: true };
    } else {
      // create a new object to avoid modifying the original options
      modules = {
        ...modules,
        exportOnlyLocals: true,
      };
    }

    return {
      ...options,
      modules,
    };
  }

  return options;
};

export async function applyBaseCSSRule(
  rule: WebpackChain.Rule,
  config: NormalizedConfig,
  context: BuilderContext,
  {
    target,
    isProd,
    isServer,
    CHAIN_ID,
    isWebWorker,
    getCompiledPath,
  }: ModifyWebpackChainUtils,
) {
  const { applyOptionsChain } = await import('@modern-js/utils');
  const browserslist = await getBrowserslistWithDefault(
    context.rootPath,
    config,
    target,
  );

  const getPostcssConfig = () => {
    const extraPlugins: AcceptedPlugin[] = [];

    const utils = {
      addPlugins(plugins: AcceptedPlugin | AcceptedPlugin[]) {
        if (Array.isArray(plugins)) {
          extraPlugins.push(...plugins);
        } else {
          extraPlugins.push(plugins);
        }
      },
    };

    const enableCssMinify = !enableExtractCSS && isProd;

    const mergedConfig = applyOptionsChain(
      {
        postcssOptions: {
          plugins: [
            require(getCompiledPath('postcss-flexbugs-fixes')),
            require(getCompiledPath('postcss-custom-properties')),
            require(getCompiledPath('postcss-initial')),
            require(getCompiledPath('postcss-page-break')),
            require(getCompiledPath('postcss-font-variant')),
            require(getCompiledPath('postcss-media-minmax')),
            require(getCompiledPath('postcss-nesting')),
            require(getCompiledPath('autoprefixer'))(
              applyOptionsChain(
                {
                  flexbox: 'no-2009',
                  overrideBrowserslist: browserslist,
                },
                config.tools.autoprefixer,
              ),
            ),
            enableCssMinify
              ? require('cssnano')(getCssnanoDefaultOptions())
              : false,
          ].filter(Boolean),
        },
        sourceMap: enableSourceMap,
      },
      // postcss-loader will modify config
      _.cloneDeep(config.tools.postcss || {}),
      utils,
    );
    if (extraPlugins.length) {
      assert('postcssOptions' in mergedConfig);
      assert('plugins' in mergedConfig.postcssOptions!);
      mergedConfig.postcssOptions.plugins!.push(...extraPlugins);
    }

    return mergedConfig as ProcessOptions & {
      postcssOptions: {
        plugins?: AcceptedPlugin[];
      };
    };
  };

  // 1. Check user config
  const enableExtractCSS = isUseCssExtract(config, target);
  const enableSourceMap = isUseCssSourceMap(config);
  const enableCSSModuleTS = Boolean(config.output.enableCssModuleTSDeclaration);
  // 2. Prepare loader options
  const extraCSSOptions: Required<CSSExtractOptions> =
    typeof config.tools.cssExtract === 'object'
      ? config.tools.cssExtract
      : {
          loaderOptions: {},
          pluginOptions: {},
        };
  const styleLoaderOptions = applyOptionsChain<StyleLoaderOptions, null>(
    {},
    config.tools.styleLoader,
  );

  // 3. Create webpack rule
  // Order: style-loader/mini-css-extract -> css-loader -> postcss-loader
  if (!isServer && !isWebWorker) {
    // use mini-css-extract-plugin loader
    if (enableExtractCSS) {
      rule
        .use(CHAIN_ID.USE.MINI_CSS_EXTRACT)
        .loader(require('mini-css-extract-plugin').loader)
        .options(extraCSSOptions.loaderOptions)
        .end();
    }
    // use style-loader
    else {
      rule
        .use(CHAIN_ID.USE.STYLE)
        .loader(require.resolve('style-loader'))
        .options(styleLoaderOptions)
        .end();
    }

    // use css-modules-typescript-loader
    if (enableCSSModuleTS) {
      rule
        .use(CHAIN_ID.USE.CSS_MODULES_TS)
        .loader(getCompiledPath('css-modules-typescript-loader'))
        .end();
    }
  } else {
    rule
      .use(CHAIN_ID.USE.IGNORE_CSS)
      .loader(path.resolve(__dirname, '../webpackLoaders/ignoreCssLoader'))
      .end();
  }

  const localIdentName =
    config.output.cssModuleLocalIdentName ||
    // Using shorter classname in production to reduce bundle size
    (isProd ? '[hash:base64:5]' : '[path][name]__[local]--[hash:base64:5]');

  const mergedCssLoaderOptions = applyOptionsChain<CSSLoaderOptions, null>(
    {
      importLoaders: 1,
      modules: {
        auto: config.output.disableCssModuleExtension
          ? isLooseCssModules
          : true,
        exportLocalsConvention: 'camelCase',
        localIdentName,
      },
      sourceMap: enableSourceMap,
    },
    config.tools.cssLoader,
    undefined,
    deepMerge,
  );
  const cssLoaderOptions = normalizeCssLoaderOptions(
    mergedCssLoaderOptions,
    isServer || isWebWorker,
  );

  rule
    .use(CHAIN_ID.USE.CSS)
    .loader(getCompiledPath('css-loader'))
    .options(cssLoaderOptions)
    .end();

  if (!isServer && !isWebWorker) {
    const postcssLoaderOptions = getPostcssConfig();
    rule
      .use(CHAIN_ID.USE.POSTCSS)
      .loader(getCompiledPath('postcss-loader'))
      .options(postcssLoaderOptions)
      .end();
  }

  // CSS imports should always be treated as sideEffects
  rule.merge({ sideEffects: true });
}

export const PluginCss = (): BuilderPlugin => {
  return {
    name: 'builder-plugin-css',
    setup(api) {
      api.modifyWebpackChain(async (chain, utils) => {
        const rule = chain.module.rule(utils.CHAIN_ID.RULE.CSS);
        const config = api.getNormalizedConfig();
        rule.test(CSS_REGEX);
        await applyBaseCSSRule(rule, config, api.context, utils);
      });
    },
  };
};
