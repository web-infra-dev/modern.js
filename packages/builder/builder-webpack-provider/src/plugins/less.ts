import { isUseCssSourceMap, LESS_REGEX } from '@modern-js/builder-shared';
import _ from '@modern-js/utils/lodash';
import type {
  BuilderPlugin,
  FileFilterUtil,
  LessLoaderOptions,
} from '../types';

export type LessLoaderUtils = {
  addExcludes: FileFilterUtil;
};

export function PluginLess(): BuilderPlugin {
  return {
    name: 'builder-plugin-less',
    setup(api) {
      api.modifyWebpackChain(async (chain, utils) => {
        const config = api.getNormalizedConfig();
        const { applyOptionsChain } = await import('@modern-js/utils');
        const { applyBaseCSSRule } = await import('./css');
        const { merge: deepMerge } = await import('@modern-js/utils/lodash');
        const getLessLoaderOptions = () => {
          const excludes: (RegExp | string)[] = [];

          const addExcludes: FileFilterUtil = items => {
            excludes.push(..._.castArray(items));
          };

          const defaultLessLoaderOptions: LessLoaderOptions = {
            lessOptions: {
              // Compat for antd or other component libraries which use math function to calculate the value
              math: 'always',
              javascriptEnabled: true,
            },
            sourceMap: isUseCssSourceMap(config),
            implementation: utils.getCompiledPath('less'),
          };
          const mergedOptions = applyOptionsChain<
            LessLoaderOptions,
            LessLoaderUtils
          >(
            defaultLessLoaderOptions,
            config.tools.less,
            { addExcludes },
            (
              defaults: LessLoaderOptions,
              userOptions: LessLoaderOptions,
            ): LessLoaderOptions => {
              return {
                ...defaults,
                ...userOptions,
                lessOptions: deepMerge(
                  defaults.lessOptions,
                  userOptions.lessOptions,
                ),
              };
            },
          );

          return {
            options: mergedOptions,
            excludes,
          };
        };

        const { options, excludes } = getLessLoaderOptions();
        const rule = chain.module
          .rule(utils.CHAIN_ID.RULE.LESS)
          .test(LESS_REGEX);

        excludes.forEach(item => {
          rule.exclude.add(item);
        });

        await applyBaseCSSRule(rule, config, api.context, utils);

        rule
          .use(utils.CHAIN_ID.USE.LESS)
          .loader(utils.getCompiledPath('less-loader'))
          .options(options);
      });
    },
  };
}
