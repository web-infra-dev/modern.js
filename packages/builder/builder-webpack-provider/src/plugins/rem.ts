import _ from '@modern-js/utils/lodash';
import type { BuilderPlugin, RemOptions, PxToRemOptions } from '../types';

const defaultOptions: RemOptions = {
  enableRuntime: true,
  rootFontSize: 50,
};

export const builderPluginRem = (): BuilderPlugin => ({
  name: 'builder-plugin-rem',

  setup(api) {
    api.modifyWebpackChain(
      async (chain, { CHAIN_ID, isServer, isWebWorker }) => {
        const {
          output: { convertToRem },
        } = api.getNormalizedConfig();

        if (!convertToRem || isServer || isWebWorker) {
          return;
        }

        const userOptions = {
          ...defaultOptions,
          ...(typeof convertToRem === 'boolean' ? {} : convertToRem),
        };

        // handle css
        const { default: PxToRemPlugin } = (await import(
          '../../compiled/postcss-pxtorem'
        )) as {
          default: (_opts: PxToRemOptions) => any;
        };

        const applyRules = [
          CHAIN_ID.RULE.CSS,
          CHAIN_ID.RULE.LESS,
          CHAIN_ID.RULE.SASS,
          CHAIN_ID.RULE.STYLUS,
        ];
        const getPxToRemPlugin = () =>
          PxToRemPlugin({
            rootValue: userOptions.rootFontSize,
            unitPrecision: 5,
            propList: ['*'],
            ..._.cloneDeep(userOptions.pxtorem || {}),
          });
        // Deep copy options to prevent unexpected behavior.
        applyRules.forEach(name => {
          chain.module.rules.has(name) &&
            chain.module
              .rule(name)
              .use(CHAIN_ID.USE.POSTCSS)
              .tap((options = {}) => ({
                ...options,
                postcssOptions: {
                  ...(options.postcssOptions || {}),
                  plugins: [
                    ...(options.postcssOptions?.plugins || []),
                    getPxToRemPlugin(),
                  ],
                },
              }));
        });

        // handle runtime (html)
        if (!userOptions.enableRuntime) {
          return;
        }

        const { AutoSetRootFontSizePlugin } = await import(
          '../webpackPlugins/AutoSetRootFontSizePlugin'
        );

        const entries = Object.keys(chain.entryPoints.entries() || {});

        chain
          .plugin(CHAIN_ID.PLUGIN.AUTO_SET_ROOT_SIZE)
          .use(AutoSetRootFontSizePlugin, [userOptions, entries]);
      },
    );
  },
});
