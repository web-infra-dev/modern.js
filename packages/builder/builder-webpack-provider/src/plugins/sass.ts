import path from 'path';
import { pathToFileURL } from 'url';
import { isUseCssSourceMap, SASS_REGEX } from '@modern-js/builder-shared';
import type { BuilderPlugin, SassLoaderOptions } from '../types';

/** fix issue about dart2js: https://github.com/dart-lang/sdk/issues/27979 */
export function patchGlobalLocation() {
  const href = `${pathToFileURL(process.cwd()).href}${path.sep}`;
  // @ts-expect-error
  global.location ||= {};
  global.location.href ||= href;
}

export function PluginSass(): BuilderPlugin {
  return {
    name: 'builder-plugin-sass',
    setup(api) {
      patchGlobalLocation();
      api.modifyWebpackChain(async (chain, utils) => {
        const config = api.getNormalizedConfig();
        const { applyOptionsChain } = await import('@modern-js/utils');
        const { applyBaseCSSRule } = await import('./css');
        const { merge: deepMerge } = await import('@modern-js/utils/lodash');

        const getSassLoaderOptions = () => {
          const excludes: RegExp[] = [];

          const addExcludes = (items: RegExp | RegExp[]) => {
            if (Array.isArray(items)) {
              excludes.push(...items);
            } else {
              excludes.push(items);
            }
          };

          const mergedOptions = applyOptionsChain<
            SassLoaderOptions,
            { addExcludes: (excludes: RegExp | RegExp[]) => void }
          >(
            {
              sourceMap: isUseCssSourceMap(config),
              implementation: utils.getCompiledPath('sass'),
            },
            config.tools.sass,
            { addExcludes },
            (
              defaults: SassLoaderOptions,
              userOptions: SassLoaderOptions,
            ): SassLoaderOptions => {
              return {
                ...defaults,
                ...userOptions,
                sassOptions: deepMerge(
                  defaults.sassOptions,
                  userOptions.sassOptions,
                ),
              };
            },
          );

          return {
            options: mergedOptions,
            excludes,
          };
        };

        const { options, excludes } = getSassLoaderOptions();
        const rule = chain.module
          .rule(utils.CHAIN_ID.RULE.SASS)
          .test(SASS_REGEX);

        excludes.forEach(item => {
          rule.exclude.add(item);
        });

        await applyBaseCSSRule(rule, config, api.context, utils);

        rule
          .use(utils.CHAIN_ID.USE.SASS)
          .loader(utils.getCompiledPath('sass-loader'))
          .options(options);
      });
    },
  };
}
