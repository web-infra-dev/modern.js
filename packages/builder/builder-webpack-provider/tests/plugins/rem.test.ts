import { describe, expect, it } from 'vitest';
import { PluginCss } from '../../src/plugins/css';
import { PluginLess } from '../../src/plugins/less';
import { PluginSass } from '../../src/plugins/sass';
import { PluginRem } from '../../src/plugins/rem';
import { createStubBuilder } from '../../src/stub';

describe('plugins/rem', () => {
  it('should not run rem plugin without config', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss(), PluginRem()],
      builderConfig: {},
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should not run rem plugin when false', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss(), PluginRem()],
      builderConfig: {
        output: {
          convertToRem: false,
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should run rem plugin with default config', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss(), PluginLess(), PluginSass(), PluginRem()],
      builderConfig: {
        output: {
          convertToRem: true,
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();

    // should add AutoSetRootFontSizePlugin and postcss-rem plugin
    expect(config).toMatchSnapshot();
  });

  it('should not run htmlPlugin with enableRuntime is false', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss(), PluginRem()],
      builderConfig: {
        output: {
          convertToRem: {
            enableRuntime: false,
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();

    expect(config.plugins?.length || 0).toBe(0);
    expect(config).toMatchSnapshot();
  });

  it('should run rem plugin with custom config', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss(), PluginRem()],
      builderConfig: {
        output: {
          convertToRem: {
            rootFontSize: 30,
            pxtorem: {
              propList: ['font-size'],
            },
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });
});
