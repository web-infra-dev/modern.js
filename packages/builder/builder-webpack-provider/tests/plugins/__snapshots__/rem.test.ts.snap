// Vitest Snapshot v1

exports[`plugins/rem > should not run htmlPlugin with enableRuntime is false 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                  {
                    "AtRule": [Function],
                    "Declaration": [Function],
                    "Once": [Function],
                    "postcssPlugin": "postcss-pxtorem",
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/rem > should not run rem plugin when false 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/rem > should not run rem plugin without config 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/rem > should run rem plugin with custom config 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                  {
                    "AtRule": [Function],
                    "Declaration": [Function],
                    "Once": [Function],
                    "postcssPlugin": "postcss-pxtorem",
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    AutoSetRootFontSizePlugin {
      "name": "AutoSetRootFontSizePlugin",
      "options": {
        "enableRuntime": true,
        "excludeEntries": [],
        "maxRootFontSize": 64,
        "pxtorem": {
          "propList": [
            "font-size",
          ],
        },
        "rootFontSize": 30,
        "rootFontSizeVariableName": "ROOT_FONT_SIZE",
        "screenWidth": 375,
        "supportLandscape": false,
        "useRootFontSizeBeyondMax": false,
        "widthQueryKey": "",
      },
      "webpackEntries": [],
    },
  ],
}
`;

exports[`plugins/rem > should run rem plugin with default config 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                  {
                    "AtRule": [Function],
                    "Declaration": [Function],
                    "Once": [Function],
                    "postcssPlugin": "postcss-pxtorem",
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
      {
        "sideEffects": true,
        "test": /\\\\\\.less\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                  {
                    "AtRule": [Function],
                    "Declaration": [Function],
                    "Once": [Function],
                    "postcssPlugin": "postcss-pxtorem",
                  },
                ],
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/less-loader",
            "options": {
              "implementation": "<ROOT>/compiled/less",
              "lessOptions": {
                "javascriptEnabled": true,
                "math": "always",
              },
              "sourceMap": true,
            },
          },
        ],
      },
      {
        "sideEffects": true,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "> 0.01%",
                      "not dead",
                      "not op_mini all",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "> 0.01%",
                        "not dead",
                        "not op_mini all",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                  {
                    "AtRule": [Function],
                    "Declaration": [Function],
                    "Once": [Function],
                    "postcssPlugin": "postcss-pxtorem",
                  },
                ],
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/compiled/sass-loader",
            "options": {
              "implementation": "<ROOT>/compiled/sass",
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": [
    AutoSetRootFontSizePlugin {
      "name": "AutoSetRootFontSizePlugin",
      "options": {
        "enableRuntime": true,
        "excludeEntries": [],
        "maxRootFontSize": 64,
        "rootFontSize": 50,
        "rootFontSizeVariableName": "ROOT_FONT_SIZE",
        "screenWidth": 375,
        "supportLandscape": false,
        "useRootFontSizeBeyondMax": false,
        "widthQueryKey": "",
      },
      "webpackEntries": [],
    },
  ],
}
`;
